### PR TITLE
VAEのコンポーネント作成

### PR DESCRIPTION
**Pull request by cehl-kurage**
## 概要
#92 #94 #95
以下を追加しました
- VAEのデコーダ用のクラス`SmallDeconvNet`
- `VAEEncoder`
- `VAEDecoder`
- `VAE`
- これらのテスト

## Submit前の確認項目

- [x] タイトルは一目でわかるようにし、説明文は PR を簡潔に説明するようにしましたか?
- [x] あなたの PR が、異なる変更を束ねたものではなく、ひとつのことだけを行うものであることを確認しましたか?
- [x] この PR で導入されたすべての変更点をリストアップしましたか?
- [x] PR を`make test`コマンドでローカルにテストしましたか?
- [x] `make format`コマンドで `pre-commit` フックを実行しましたか?

### CODE REVIEW

**Comment by cehl-kurage** [comment]
実装ミスにつき修正中

File: src/models/components/small_deconv_net.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        256)に合わせるため、2回目のConvTranspose2dにoutput_padding=1を追加しています。
+
+        Args:
+            height (int): 再構成画像の高さ
+            width (int): 再構成画像の横幅
+            channels (int): 再構成画像のチャネル数
+            dim_in (int): 潜在変数ベクトルの次元数
+            positional_bias (bool): 最後にバイアス項を足すかどうか
+            nl (Callable): 活性化関数
+        """
```
Comment (Geson-anko) [inline src/models/components/small_deconv_net.py:28]: 英語にしてください。

File: src/models/components/small_deconv_net.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            output_padding=1,
+        )  # 追加部分
+        self.deconv3 = nn.ConvTranspose2d(
+            64, 3, kernel_size=self.kernel_sizes[2], stride=self.strides[2], padding=self.paddings[2]
+        )
+        self.bias = nn.Parameter(torch.zeros(channels, height, width), requires_grad=True) if positional_bias else None
+        self.nl = nl
+
+    @property
+    def init_output_size(self):
```
Comment (Geson-anko) [inline src/models/components/small_deconv_net.py:59]: こういう複雑な事をするときはdocstringに何を行っているか概要を記述してください。できれば `_compute_input_shape`メソッドも **何のための計算なのか** 明記してください。

File: src/models/components/small_deconv_net.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        output_size = (self.height, self.width)
+        for kernel_size, stride, padding in zip(self.kernel_sizes[::-1], self.strides[::-1], self.paddings[::-1]):
+            output_size = tuple(map(self._compute_input_shape, output_size, kernel_size, stride, padding))
+        return output_size
+
+    def _compute_input_shape(
```
Comment (Geson-anko) [inline src/models/components/small_deconv_net.py:65]: 多分PyTorchのConvTranspose2dの、outputの計算式から逆関数を算出してやってるのだと思いますが、参照先があればそれを記載すると良いと思います。
https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html?highlight=convtranspose2d#torch.nn.ConvTranspose2d

ちなみにConvTransposeのサイズ I/Oの計算式は次のようになり、

$$
L_{out} = (L_{in} - 1) \times \text{stride} - 2 \times \text{padding} + \text{dilation} \times (\text{kernel size} - 1) + \text{output padding} + 1
$$

逆関数は output paddingを加味してこのようになります。

$$
L_{out} = \left\lfloor\frac{L_{in} + 2 \times \text{padding} - \text{dilation} \times (\text{kernel size} - 1) - 1 - \text{output padding}}{\text{stride}} + 1\right\rfloor
$$

File: tests/models/components/test_small_deconv_net.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        """batch,
+            height,
+            width,
+            channels,
+            dim_in,
+            positional_bias,
+            nl""",
+        [
+            (8, 256, 256, 3, 256, False, torch.nn.LeakyReLU()),
+            (1, 256, 256, 3, 128, True, torch.nn.LeakyReLU(negative_slope=0.2)),
```
Comment (Geson-anko) [inline tests/models/components/test_small_deconv_net.py:33]: 画像サイズが 128, 256, 512の場合でもテストしてください。

**Review by Geson-anko — CHANGES_REQUESTED** [review]
@cehl-kurage 実装お疲れ様です。ConvTransposeの入出力計算面倒ですよね。難しいようでしたら私も手伝います。

`encode`メソッドや`decode`メソッドはそれぞれに`forward`として書いちゃっていいと思います。何か別な機能を実装する予定があれば、メソッドを分けた方がいいのですが、現状ただのエイリアスになっているので冗長かと

File: src/models/components/observation_encoder/vae.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+from torch import Tensor
+from torch.distributions.normal import Normal
+
+from ..small_conv_net import SmallConvNet
+from ..small_deconv_net import SmallDeconvNet
+from .observation_encoder import ObservationEncoder
+
+
+class VAEEncoder(ObservationEncoder):
+    def __init__(self, small_conv_net: SmallConvNet, min_stddev=0.0):
```
[the remark this answers is not in the document]
Comment (cehl-kurage) [reply src/models/components/observation_encoder/vae.py:14]: ここが根拠でしたが、そのとおりなので修正します。  
https://github.com/openai/large-scale-curiosity/blob/master/auxiliary_tasks.py#L99C5-L99C5

File: src/models/components/small_deconv_net.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            dim_in (int): 潜在変数ベクトルの次元数
+            positional_bias (bool): 最後にバイアス項を足すかどうか
+            nl (Callable): 活性化関数
+        """
+        super().__init__()
+        self.height = height
+        self.width = width
+        self.channels = channels
+        self.kernel_sizes = ((4, 4), (8, 8), (8, 8))
+        self.strides = ((2, 2), (2, 2), (3, 3))
```
Comment (Geson-anko) [inline src/models/components/small_deconv_net.py:34]: 多分 256に合わせるのが大変になってる場所として、strideが3なのが問題だと思う。1->2->4とか、2->2->4とかがいいのでは？正直ここら辺の実装の違いは仕方ないと思う。

Comment (cehl-kurage) [reply src/models/components/small_deconv_net.py:34]: そうですね。そこで汎用性もなくなってると思うので、それでよければできると思います。  
承知しました。

File: tests/models/components/observation_encoder/test_vae.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+    def test__init__(self, small_conv_net):
+        mod = Encoder(small_conv_net)
+        assert mod.conv_net is small_conv_net
+
+    def test_forward(self, small_conv_net):
+        mod = Encoder(small_conv_net)
+        x = torch.randn(8, CHANNELS, WIDTH, HEIGHT)
+        z_dist = mod(x)
+        assert isinstance(z_dist, torch.distributions.Normal)
```
Comment (Geson-anko) [inline tests/models/components/observation_encoder/test_vae.py:30]:
```suggestion
        assert isinstance(z_dist, torch.distributions.Normal)
        assert z_dist.rsample().shape == (8, DIM_EMBED)
```

**Review by Geson-anko — APPROVED** [review]
@cehl-kurage OK! コンフリクト直しておきました。若干の修正はsuggestionからcommitしておきました。

File: tests/models/components/test_small_deconv_net.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+from src.models.components.small_deconv_net import SmallDeconvNet as cls
+
+
+class TestSmallDeconvNet:
+    def test__init__(self):
+        mod = cls(256, 256, 3, 256, True, torch.nn.ReLU())
+        assert mod.channels == 3
+        assert mod.fc_init.in_features == 256
+        assert mod.bias.shape == (3, 256, 256)
+        assert mod.nl == torch.nn.ReLU()
```
Comment (Geson-anko) [inline tests/models/components/test_small_deconv_net.py:13]: 等価表現は使えませんので、`isinstance`を用いましょう。
```suggestion
        assert isinstance(mod.nl, torch.nn.ReLU)
```

**Comment by Geson-anko** [comment]
@cehl-kurage テスト落ちてるんで修正してくださいー
